### PR TITLE
Add option to disable jump draw

### PR DESCRIPTION
**Pull request by shahril96**
Changes made:

1. Added option to disable jump draw

image

2. Fixed possible crashing when accessing show_addresses_ function 

3. Dock registers badge at the left side of line1

image

Any comments is really appreciated. :)

### CODE REVIEW

File: src/widgets/QDisassemblyView.cpp [hunk 1]
```diff
@@ -1370,7 +1370,8 @@ void QDisassemblyView::drawJumpArrows(QPainter &painter, const DrawingContext *c
 		// if direct jmp is selected, then draw arrow in red
 		if (is_unconditional_jump(instructions_[jump_arrow.src_line]) && 
 			(ctx->selected_line == jump_arrow.src_line || 
-			(ctx->selected_line == jump_arrow.dst_line && show_addresses_[jump_arrow.dst_line] != current_address_ ))) {
+			(ctx->selected_line == jump_arrow.dst_line && 
+			(jump_arrow.dst_in_viewport && show_addresses_[jump_arrow.dst_line] != current_address_ )))) {
```
Comment (10110111) [inline src/widgets/QDisassemblyView.cpp]: Nitpick: the additional parentheses are superfluous here, and they do add to clutter on the closing side: `))))`. I'd suggest removing this pair, retaining the logically related conditions on the single line as they are now.

**Comment by 10110111** [comment]
When I set _Show jump arrows_ and unset _Show when registers are pointing to a visible address_, the space allocated appears to be large enough for both, although only one thing is enabled.

File: src/widgets/QDisassemblyView.cpp (hunk 1 above)
Comment (shahril96) [reply src/widgets/QDisassemblyView.cpp]: Fixed

**Comment by shahril96** [comment]
@10110111, that is an intended behavior, as register badge is drawn inside jump arrow area (at left side of first line)

**Comment by 10110111** [comment]
It's not drawn when it's disabled, is it?

**Comment by shahril96** [comment]
Peek 2019-11-02 03-48

Register badge is drawn inside jump arrow area, so there is no effect on area allocated if either the badge is drawn or not.

Jump arrow area is always going to be initially as this size (https://github.com/eteran/edb-debugger/pull/731/files#diff-4af38e37cb69a85d76c168f3230a7adeL1650)

Unless I got you wrong 😅

**Comment by 10110111** [comment]
Ah, so you made it more economical... But don't you think it's not very nice when the arrows go over the badges, like in the following screenshot?
Screenshot - 011119 - 23:09:30

**Comment by shahril96** [comment]
@10110111 that is a good catch (doesn't appear in my test 😅)

thank you! will fix it soon

File: src/widgets/QDisassemblyView.cpp [hunk 2]
```diff
@@ -1647,11 +1676,19 @@ int QDisassemblyView::line0() const {
 // Desc:
 //------------------------------------------------------------------------------
 int QDisassemblyView::line1() const {
-	if(line1_ == 0) {
+
+	if(!edb::v1::config().show_jump_arrow) {
+
+		// allocate space for register badge
+		// 5 (maximum register name length i can find off) + overhead
```
Comment (10110111) [inline src/widgets/QDisassemblyView.cpp]: Which of the registers has 5-char name? `MXCSR`? `YMM12`? I would be surprised if they were able to point anywhere ;)
Only GPRs are used for the badges, see the beginning of `drawRegiserBadges()`.

Comment (shahril96) [reply src/widgets/QDisassemblyView.cpp]: Fixed

**Comment by shahril96** [comment]
@10110111 latest commits fixed issues that you pointed out

File: src/widgets/QDisassemblyView.cpp [hunk 3]
```diff
@@ -1283,12 +1283,35 @@ void QDisassemblyView::drawJumpArrows(QPainter &painter, const DrawingContext *c
 		return a.distance < b.distance;
 	});
 
+	auto isLineOverlap = [&](int line1_head, int line1_tail, int line2_head, int line2_tail, bool edge_overlap = true) -> bool {
```
Comment (10110111) [inline src/widgets/QDisassemblyView.cpp]: You are not using this default `= true`, so no point in it.

File: src/widgets/QDisassemblyView.cpp [hunk 4]
```diff
@@ -1689,8 +1724,8 @@ int QDisassemblyView::line1() const {
 	if(!edb::v1::config().show_jump_arrow) {
 
 		// allocate space for register badge
-		// 5 (maximum register name length i can find off) + overhead
-		return (edb::v1::config().show_register_badges) ? (6 * font_width_ + font_width_/2) : 0;
+		// 4 (maximum register name for GPR) + overhead
+		return (edb::v1::config().show_register_badges) ? (5 * font_width_ + font_width_/2) : 0;
```
Comment (10110111) [inline src/widgets/QDisassemblyView.cpp]: Why parentheses around single variable?

**Comment by 10110111** [comment]
In other respects, this looks fine to me now.

**Comment by shahril96** [comment]
Pushed fixes as mentioned by @10110111

File: src/widgets/QDisassemblyView.cpp (hunk 4 above)
Comment (shahril96) [reply src/widgets/QDisassemblyView.cpp]: fixes

File: src/widgets/QDisassemblyView.cpp (hunk 3 above)
Comment (shahril96) [reply src/widgets/QDisassemblyView.cpp]: fixed

**Comment by eteran** [comment]
Looks good to me! Merging!

@10110111 Thanks for the code review, @shahril96 thanks for this awesome contribution!

**Pull request merged**